[BUG] Reading the iceberg metadata failed with java.lang.ArrayIndexOutOfBoundsException

Reading the iceberg metadata failed with java.lang.ArrayIndexOutOfBoundsException.
Such as
```
spark-shell --packages org.apache.iceberg:iceberg-spark-runtime-3.2_2.12:0.13.0 \
    --conf spark.sql.session.timeZone=UTC \
    --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions \
    --conf spark.sql.catalog.spark_catalog=org.apache.iceberg.spark.SparkSessionCatalog \
    --conf spark.sql.catalog.spark_catalog.type=hive

spark.sql("""CREATE TABLE spark_catalog.default.demo2 (id bigint, data string) USING iceberg;""")
spark.sql("""INSERT INTO spark_catalog.default.demo2 VALUES (1, 'a'), (2, 'b'), (3, 'c');""")

spark.read.format("iceberg").load("spark_catalog.default.demo2.files").count
```

Stacktrace:
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
  at com.nvidia.spark.rapids.GpuRowToColumnConverter.$anonfun$convert$1(GpuRowToColumnarExec.scala:48)
  at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
  at com.nvidia.spark.rapids.GpuRowToColumnConverter.convert(GpuRowToColumnarExec.scala:47)
  at com.nvidia.spark.rapids.RowToColumnarIterator.$anonfun$buildBatch$1(GpuRowToColumnarExec.scala:639)
  at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
  at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
  at com.nvidia.spark.rapids.RowToColumnarIterator.withResource(GpuRowToColumnarExec.scala:588)
  at com.nvidia.spark.rapids.RowToColumnarIterator.buildBatch(GpuRowToColumnarExec.scala:615)
  at com.nvidia.spark.rapids.RowToColumnarIterator.next(GpuRowToColumnarExec.scala:611)
  at com.nvidia.spark.rapids.RowToColumnarIterator.next(GpuRowToColumnarExec.scala:588)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.aggregateInputBatches(aggregate.scala:283)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$next$2(aggregate.scala:238)
  at scala.Option.getOrElse(Option.scala:189)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:235)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:181)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.aggregateInputBatches(aggregate.scala:283)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$next$2(aggregate.scala:238)
  at scala.Option.getOrElse(Option.scala:189)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:235)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:181)
  at com.nvidia.spark.rapids.ColumnarToRowIterator.$anonfun$fetchNextBatch$2(GpuColumnarToRowExec.scala:241)
  at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
  at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
  at com.nvidia.spark.rapids.ColumnarToRowIterator.withResource(GpuColumnarToRowExec.scala:187)
  at com.nvidia.spark.rapids.ColumnarToRowIterator.fetchNextBatch(GpuColumnarToRowExec.scala:238)
  at com.nvidia.spark.rapids.ColumnarToRowIterator.loadNextBatch(GpuColumnarToRowExec.scala:215)
  at com.nvidia.spark.rapids.ColumnarToRowIterator.hasNext(GpuColumnarToRowExec.scala:255)
  at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:349)
  at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
  at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
  at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
  at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
  at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
  at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
  at org.apache.spark.scheduler.Task.run(Task.scala:131)
  at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
  at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)
  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:750)
```

--- a/integration_tests/src/main/python/iceberg_test.py
+++ b/integration_tests/src/main/python/iceberg_test.py
@@ -14,7 +14,7 @@
 
 import pytest
 
-from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_fallback_collect, assert_py4j_exception
+from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_and_cpu_row_counts_equal, assert_gpu_fallback_collect, assert_py4j_exception
 from data_gen import *
 from marks import allow_non_gpu, iceberg, ignore_order
 from spark_session import is_before_spark_320, is_databricks_runtime, with_cpu_session, with_gpu_session
@@ -232,6 +232,28 @@ def test_iceberg_read_metadata_fallback(spark_tmp_table_factory):
         assert_gpu_fallback_collect(
             lambda spark : spark.read.format("iceberg").load("default.{}.{}".format(table, subtable)),
             "BatchScanExec")
+
+@iceberg
+# Some metadata files have types that are not supported on the GPU yet (e.g.: BinaryType)
+@allow_non_gpu("BatchScanExec", "ProjectExec")
+def test_iceberg_read_metadata_count(spark_tmp_table_factory):
+    table = spark_tmp_table_factory.get()
+    tmpview = spark_tmp_table_factory.get()
+    def setup_iceberg_table(spark):
+        df = binary_op_df(spark, long_gen)
+        df.createOrReplaceTempView(tmpview)
+        spark.sql("CREATE TABLE {} USING ICEBERG ".format(table) + \
+                  "AS SELECT * FROM {}".format(tmpview))
+        df = binary_op_df(spark, long_gen, seed=1)
+        df.createOrReplaceTempView(tmpview)
+        spark.sql("INSERT INTO {} ".format(table) + \
+                  "SELECT * FROM {}".format(tmpview))
+    with_cpu_session(setup_iceberg_table)
+    for subtable in ["all_data_files", "all_manifests", "files", "history",
+                     "manifests", "partitions", "snapshots"]:
+        # SQL does not have syntax to read table metadata
+        assert_gpu_and_cpu_row_counts_equal(
+            lambda spark : spark.read.format("iceberg").load("default.{}.{}".format(table, subtable)))
 
 @iceberg
 @ignore_order(local=True) # Iceberg plans with a thread pool and is not deterministic in file ordering

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRowToColumnarExec.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRowToColumnarExec.scala
@@ -44,7 +44,7 @@ private class GpuRowToColumnConverter(schema: StructType) extends Serializable w
    */
   final def convert(row: InternalRow, builders: GpuColumnarBatchBuilder): Double = {
     var bytes: Double = 0
-    for (idx <- 0 until row.numFields) {
+    for (idx <- schema.fields.indices) {
       bytes += converters(idx).append(row, idx, builders.builder(idx))
     }
     bytes

